cmd/bad-key-revoker/main.go

```diff
@@ -461,29 +472,74 @@ func main() {
 	cmd.FailOnError(err, fmt.Sprintf("failed to parse email template %q: %s", config.BadKeyRevoker.Mailer.EmailTemplate, err))
 
 	bkr := &badKeyRevoker{
-		dbMap:           dbMap,
-		maxRevocations:  config.BadKeyRevoker.MaximumRevocations,
-		serialBatchSize: config.BadKeyRevoker.FindCertificatesBatchSize,
-		raClient:        rac,
-		mailer:          mailClient,
-		emailSubject:    config.BadKeyRevoker.Mailer.EmailSubject,
-		emailTemplate:   emailTemplate,
-		logger:          logger,
-		clk:             clk,
+		dbMap:               dbMap,
+		maxRevocations:      config.BadKeyRevoker.MaximumRevocations,
+		serialBatchSize:     config.BadKeyRevoker.FindCertificatesBatchSize,
+		raClient:            rac,
+		mailer:              mailClient,
+		emailSubject:        config.BadKeyRevoker.Mailer.EmailSubject,
+		emailTemplate:       emailTemplate,
+		logger:              logger,
+		clk:                 clk,
+		backoffDurationMax:  config.BadKeyRevoker.BackoffDurationMax.Duration,
+		backoffDurationBase: config.BadKeyRevoker.Interval.Duration,
+		backoffFactor:       1.3,
+	}
+
+	// If `BackoffDurationMax` was not set via the config, set it to 60
+	// seconds. This will avoid a tight loop on error but not be an
+	// excessive delay if the config value was not deliberately set.
+	if config.BadKeyRevoker.BackoffDurationMax.Duration == 0 {
+		bkr.backoffDurationMax = time.Second * 60
 	}
+
+	// If `Interval` is not set in the config then set
+	// `bkr.backoffDurationBase` to a default 1 second.
+	if config.BadKeyRevoker.Interval.Duration == 0 {
+		bkr.backoffDurationBase = time.Second
+	}
+
+	// Run bad-key-revoker in a loop. Backoff if no work or errors.
 	for {
 		noWork, err := bkr.invoke()
 		if err != nil {
 			keysProcessed.WithLabelValues("error").Inc()
 			logger.AuditErrf("failed to process blockedKeys row: %s", err)
+			// Increase backoff duration.
+			bkr.backoffTick()
+			logger.Infof("backoff on error, trying again in %f seconds.", bkr.backoffDuration.Seconds())
+			// Sleep for backoff duration.
+			bkr.clk.Sleep(bkr.backoffDuration)
```

Comment:
Both here and below we do the same three calls: increase the tick, print a message, then sleep. What if we instead changed the flow to be 1) print a message; 2) sleep; 3) increase the tick for next time? That way we could combine the second and third steps into a single function call (outlined below).

Reply:
Done (more info in other comment)